### PR TITLE
TIP-1283: Create ProductModelDescendantsAndAncestorsIndexer

### PR DESCRIPTION
**Pull request by nmarniesse**
**Description (for Contributor and Core Developer)**

[TIP-1283](https://akeneo.atlassian.net/browse/TIP-1283) Create ProductModelDescendantsAndAncestorsIndexer

- add GetAncestorAndDescendantProductModelCodes query
- add GetDescendantVariantProductIdentifiers query
- add ProductModelDescendantsAndAncestorsIndexer

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | OK
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelDescendantsAndAncestorsIndexer.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+/**
+ * Indexer responsible for the indexing of the product models, product models ancestors
+ * and all children (product models and variant products)
+ *
+ * @author    Nicolas Marniesse <nicolas.marniesse@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class ProductModelDescendantsAndAncestorsIndexer
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelDescendantsAndAncestorsIndexer.php]: `ProductModelIndexerWithDescendantAndAncestors` ?. Btw, ancestors is maybe good, but not very consistent with descendant.

Parent/children, descendant/ascendant, ancestors/? 


We should ask to a PM.

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelDescendantsAndAncestorsIndexer.php [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        $this->getAncestorAndDescendantProductModelCodes = $getAncestorAndDescendantProductModelCodes;
+    }
+
+    /**
+     * Indexes the given product models with children (subtrees made of product variants and product models).
+     *
+     * @param string[] $productModelCodes
+     * @param array    $options
+     */
+    public function indexFromProductModelCodes(array $productModelCodes, array $options = []): void
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelDescendantsAndAncestorsIndexer.php]: could you remove the option please? (I think we should do it in the other indexer too).

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelDescendantsAndAncestorsIndexer.php [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        ;
+        $this->productModelIndexer->indexFromProductModelCodes(
+            array_unique(array_merge($productModelCodes, $ancestorAndDescendantsProductModelCodes)),
+            $options
+        );
+
+        $variantProductIdentifiers = $this->getDescendantVariantProductIdentifiers->fromProductModelCodes(
+            $productModelCodes
+        );
+        if (!empty($variantProductIdentifiers)) {
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelDescendantsAndAncestorsIndexer.php]: Please batch it. If you have 100 children for 100 product models, it's 10 000 products :)

`$variantProductIdentifiersByBatch = array_chunk($variantProductIdentifiers, 1000)`

File: tests/back/Pim/Enrichment/Integration/Product/Query/Sql/GetAncestorAndDescendantProductModelCodesIntegration.php [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+                'values' => [
+                    'a_simple_select' => [
+                        ['locale' => null, 'scope' => null, 'data' => 'optionA'],
+                    ],
+                ],
+            ]
+        );
+    }
+
+    protected function getAncestorAndDescendantProductModelCodes(): GetAncestorAndDescendantProductModelCodes
```
Comment (ahocquard) [inline tests/back/Pim/Enrichment/Integration/Product/Query/Sql/GetAncestorAndDescendantProductModelCodesIntegration.php]: generally, we put protected at the end of the function (public first, then protected). Exception for the setup. I think it's a good politic, but for the tests it is less important.

File: tests/back/Pim/Enrichment/Integration/Product/Query/Sql/GetAncestorAndDescendantProductModelCodesIntegration.php [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        Assert::assertEqualsCanonicalizing(
+            [],
+            $this->getAncestorAndDescendantProductModelCodes()->fromProductModelCodes([])
+        );
+    }
+
+    private function createFamilyVariant(array $data = []): FamilyVariantInterface
+    {
+        $family = $this->get('pim_catalog.factory.family_variant')->create();
+        $this->get('pim_catalog.updater.family_variant')->update($family, $data);
```
Comment (ahocquard) [inline tests/back/Pim/Enrichment/Integration/Product/Query/Sql/GetAncestorAndDescendantProductModelCodesIntegration.php]: missing validation

```
$errors = $this->get('validator')->validate($productModel);
Assert::assertCount(0, $errors);
```

File: tests/back/Pim/Enrichment/Integration/Product/Query/Sql/GetAncestorAndDescendantProductModelCodesIntegration.php [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+        return $family;
+    }
+
+    private function createProductModel(array $data = []): ProductModelInterface
+    {
+        $productModel = $this->get('pim_catalog.factory.product_model')->create();
+        $this->get('pim_catalog.updater.product_model')->update($productModel, $data);
+
+        $errors = $this->get('pim_catalog.validator.product')->validate($productModel);
```
Comment (ahocquard) [inline tests/back/Pim/Enrichment/Integration/Product/Query/Sql/GetAncestorAndDescendantProductModelCodesIntegration.php]: `Assert::assertCount(0, $errors);` is enough in my opinion

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelDescendantsAndAncestorsIndexer.php (hunk 3 above)
Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelDescendantsAndAncestorsIndexer.php]: I think the batch should be in ProductIndexer and ProductModelIndexer. No matter who call them, batch implems will be always here.  
And if someday we have a new ProductIndexer that indexes in *whatever* solution, the batch will change according to the new way to index. So I guess it's the responsability of ProductIndexer and ProductModelIndexer

Comment (ahocquard) [reply src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelDescendantsAndAncestorsIndexer.php]: > And if someday we have a new ProductIndexer that indexes in whatever solution, the batch will change according to the new way to index

You can see it as pros, but many time it was a cons in the PIM (datagrid to batch pages, api to batch products, etc). You are creating coupling, and from our experience on the PIM, global configuration is painful as soon as you want:
- to have a different behavior with a different batch size
- to fix something at the support level with more control over it

But I don't care that much and it's GTM for me with both solutions :)

File: tests/back/Pim/Enrichment/Integration/Product/Query/Sql/GetAncestorAndDescendantProductModelCodesIntegration.php [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    protected function getConfiguration()
+    {
+        return $this->catalog->useTechnicalCatalog();
+    }
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->createFamilyVariant(
```
Comment (mmetayer) [inline tests/back/Pim/Enrichment/Integration/Product/Query/Sql/GetAncestorAndDescendantProductModelCodesIntegration.php]: just a remark: there is already suche a family variant in the technical catalog (familyVariantA1)

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelDescendantsAndAncestorsIndexer.php (hunk 1 above)
Comment (mmetayer) [reply src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelDescendantsAndAncestorsIndexer.php]: IMHO "ancestor" is more explicit (and I guess more used than "ascendant" in English language, as it is in French)

Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelDescendantsAndAncestorsIndexer.php]: Well, the indexer for product and its ancestors is `ProductAndAncestorsIndexer` (TIP-1281). That's why it's `ProductModelDescendantsAndAncestorsIndexer` here. If we want to be consistent we should discuss about that too. Seems we should have discussed about that before to start :(

Comment (ahocquard) [reply src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelDescendantsAndAncestorsIndexer.php]: Anyway, let's merge it this way, renaming is quite easy with an IDE in 2019 :) But we should ask a PM to have the same vocabulary everywhere.

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelDescendantsAndAncestorsIndexer.php (hunk 3 above)
Comment (ahocquard) [reply src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Indexer/ProductModelDescendantsAndAncestorsIndexer.php]: @nmarniesse ok let's batch in ProductIndexer, I changed my mind during the night :smile: (in another PR please)

File: tests/back/Pim/Enrichment/Integration/Product/Query/Sql/GetAncestorAndDescendantProductModelCodesIntegration.php (hunk 7 above)
Comment (nmarniesse) [reply tests/back/Pim/Enrichment/Integration/Product/Query/Sql/GetAncestorAndDescendantProductModelCodesIntegration.php]: Thanks :+1:

**Pull request merged**